work-queue/sa-work-queue/src/threadpool.rs

```diff
@@ -0,0 +1,248 @@
+use futures::StreamExt;
+use lapin::{
+	message::Delivery,
+	options::{BasicAckOptions, BasicConsumeOptions, BasicNackOptions, BasicQosOptions},
+	types::FieldTable,
+	Connection, ConnectionProperties, Consumer,
+};
+use threadpool::ThreadPool;
+
+use crate::{error::*, job::BackgroundJob, runner::Event};
+
+thread_local!(static CONSUMER: ConsumerHandle = Default::default());
+
+#[derive(PartialEq, Clone, Debug)]
+struct QueueOpts {
+	queue_name: String,
+	addr: String,
+	prefetch: u16,
+}
+
+impl Default for QueueOpts {
+	fn default() -> Self {
+		Self { queue_name: "TASK_QUEUE".to_string(), addr: "amqp://localhost:5672".to_string(), prefetch: 1 }
+	}
+}
+
+impl QueueOpts {
+	fn create_connection(&self) -> Result<Connection, Error> {
+		Ok(Connection::connect(&self.addr, ConnectionProperties::default().with_async_std()).wait()?)
+	}
+}
+
+#[derive(Default)]
+pub struct Builder {
+	opts: QueueOpts,
+	threads: Option<usize>,
+	name: Option<String>,
+}
+
+impl Builder {
+	pub fn queue_name<S: AsRef<str>>(mut self, name: S) -> Self {
+		self.opts.queue_name = name.as_ref().to_string();
+		self
+	}
+
+	pub fn addr<S: AsRef<str>>(mut self, addr: S) -> Self {
+		self.opts.addr = addr.as_ref().to_string();
+		self
+	}
+
+	pub fn prefetch(mut self, prefetch: u16) -> Self {
+		self.opts.prefetch = prefetch;
+		self
+	}
+
+	pub fn threads(mut self, threads: usize) -> Self {
+		self.threads = Some(threads);
+		self
+	}
+
+	pub fn name<S: AsRef<str>>(mut self, name: S) -> Self {
+		self.name = Some(name.as_ref().to_string());
+		self
+	}
+
+	pub fn build(self) -> Result<ThreadPoolMq, Error> {
+		let conn = Arc::new(self.opts.create_connection()?);
+		let pool = ThreadPool::with_name(
+			self.name.unwrap_or_else(|| "work-queue".into()),
+			self.threads.unwrap_or_else(num_cpus::get),
+		);
+		let (tx, rx) = flume::bounded(pool.max_count());
+
+		Ok(ThreadPoolMq { conn, tx, rx, pool, queue_opts: Arc::new(self.opts) })
+	}
+}
+
+pub struct ThreadPoolMq {
+	conn: Arc<Connection>,
+	queue_opts: Arc<QueueOpts>,
+	pool: ThreadPool,
+	tx: Sender<Event>,
+	rx: Receiver<Event>,
+}
+
+impl ThreadPoolMq {
+	pub fn builder() -> Builder {
+		Default::default()
+	}
+
+	/// Execute a job on this threadpool.
+	/// Automatically advances RabbitMq queue and feeds
+	/// the payload in to the predicate `F`.
+	pub fn execute<F>(&self, job: F)
+	where
+		F: Send + 'static + FnOnce(BackgroundJob) -> Result<(), PerformError>,
+	{
+		let conn = self.conn.clone();
+		let tx = self.tx.clone();
+		let queue_opts = self.queue_opts.clone();
+		self.pool.execute(move || {
+			if let Err(e) = run_job(&conn, &queue_opts, tx, job) {
+				log::error!("{}", e);
+			}
+		})
+	}
+
+	pub fn max_count(&self) -> usize {
+		self.pool.max_count()
+	}
+
+	pub fn active_count(&self) -> usize {
+		self.pool.active_count()
+	}
+
+	pub fn queued_count(&self) -> usize {
+		self.pool.queued_count()
+	}
+
+	// TODO: could wrap this so we're not exposing underlying details and just returning a raw
```

Comment:
While designing this part I thought it might be good to wrap `Receiver` in a different struct in case we want to change `Receiver` to something else (I think I had doubts about the design at this point), but I think it's OK the way it is.